lib/matplotlib/backend_bases.py

```diff
@@ -2984,65 +2971,53 @@ def drag_pan(self, event):
-            elif event.key == "y":
-                x1, x2 = a.bbox.intervalx
-            self.draw_rubberband(event, x1, y1, x2, y2)
+        start_xy = self._zoom_info["start_xy"]
+        ax = self._zoom_info["axes"][0]
+        (x1, y1), (x2, y2) = np.clip(
+            [start_xy, [event.x, event.y]], ax.bbox.min, ax.bbox.max)
+        if event.key == "x":
+            y1, y2 = ax.bbox.intervaly
+        elif event.key == "y":
+            x1, x2 = ax.bbox.intervalx
+        self.draw_rubberband(event, x1, y1, x2, y2)
 
     def release_zoom(self, event):
         """Callback for mouse button release in zoom to rect mode."""
-        if self._id_zoom is not None:
-            self.canvas.mpl_disconnect(self._id_zoom)
-        self._id_zoom = None
-
-        self.remove_rubberband()
-
-        if not self._xypress:
+        if self._zoom_info is None:
             return
 
-        last_a = []
+        # We don't check the event button here, so that zooms can be cancelled
+        # by (pressing and) releasing another mouse button.
+        self.canvas.mpl_disconnect(self._zoom_info["cid"])
+        self.remove_rubberband()
 
-        for lastx, lasty, a in self._xypress:
+        start_x, start_y = self._zoom_info["start_xy"]
+        for i, ax in enumerate(self._zoom_info["axes"]):
             x, y = event.x, event.y
             # ignore singular clicks - 5 pixels is a threshold
             # allows the user to "cancel" a zoom action
             # by zooming by less than 5 pixels
-            if ((abs(x - lastx) < 5 and event.key != "y") or
-                    (abs(y - lasty) < 5 and event.key != "x")):
+            if ((abs(x - start_x) < 5 and event.key != "y") or
+                    (abs(y - start_y) < 5 and event.key != "x")):
                 self._xypress = None
                 self.release(event)
                 self.draw()
                 return
 
-            # detect twinx, twiny axes and avoid double zooming
-            twinx, twiny = False, False
-            if last_a:
-                for la in last_a:
-                    if a.get_shared_x_axes().joined(a, la):
-                        twinx = True
-                    if a.get_shared_y_axes().joined(a, la):
-                        twiny = True
-            last_a.append(a)
-
-            if self._button_pressed == 1:
-                direction = 'in'
-            elif self._button_pressed == 3:
-                direction = 'out'
-            else:
-                continue
+            # Detect whether this axes is twinned with an earlier axes in the
+            # list of zoomed axes, to avoid double zooming.
+            twinx = any(ax.get_shared_x_axes().joined(ax, prev)
+                        for prev in self._zoom_info["axes"][:i])
+            twiny = any(ax.get_shared_y_axes().joined(ax, prev)
+                        for prev in self._zoom_info["axes"][:i])
+
+            direction = {1: "in", 3: "out"}[self._zoom_info["start_button"]]
```

Comment:
Minor nitpicking suggestion: making a dictionary seems like overkill.  Simpler alternative:
```python
direction = "in" if self._zoom_info["start_button"] == 1 else "out"
```
Actually, now I'm wondering why this line is inside the loop over axes.

Reply:
I switched to just resolving that when creating zoom_info in press_zoom.